File: apps/android/aura/AuraApp.kt
// aura — Android (Jetpack Compose, Material 3) structural skeleton. Translates the
// shared design system (docs/DESIGN_SYSTEM.md) to Material patterns: Scaffold +
// NavigationBar (NavigationRail on expanded width), ModalBottomSheet check-in, fixed
// brand dark theme (dynamic color OFF), reduce-motion-aware orb. Compiles against a
// Kotlin port of the engine core (docs/PLATFORMS.md, strategy 1); `AuraEngine` is the
// interface that port implements.

package app.aura

import androidx.compose.animation.core.*
import androidx.compose.foundation.Canvas
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import java.util.Date

// ── Design tokens ────────────────────────────────────────────────────────────
object Tokens {
    val ink = Color(0xFF09080F)
    val card = Color.White.copy(alpha = 0.035f)
    val line = Color.White.copy(alpha = 0.09f)
    val text = Color(0xFFEDEAF6)
    val textDim = Color(0xFFB4AFC6)
    val textFaint = Color(0xFF7C7791)
    // Bundled fonts; FontFamily(Font(R.font.instrument_serif)) etc. in the real project.
    val serif = FontFamily.Serif
    val grotesk = FontFamily.SansSerif
    val sans = FontFamily.SansSerif
}

// ── Domain models (mirror @aura/engine types.ts) ─────────────────────────────
enum class Energy(val label: String, val color: Color) {
    MAIN("Main Character", Color(0xFFFFD070)), FEEL("Big Feelings", Color(0xFF8FB7FF)),
    FIRE("Fired Up", Color(0xFFFF6E58)), MIND("Busy Mind", Color(0xFF5FE0C0)),
    GROW("Green Light", Color(0xFF7ED69B)), LOVE("Soft Spot", Color(0xFFF49CC9)),
    BUILD("Heavy Lifting", Color(0xFF8E93C8)), CRAVE("Never Enough", Color(0xFFAE8FE6)),
    LET("Letting Go", Color(0xFFA6ABB8))
}

data class Reading(val headline: String, val gift: String, val trap: String, val move: String,
                   val watch: String, val remedy: String, val energy: Energy, val passing: Energy)
data class DailyBundle(val major: Energy, val passing: Energy, val todayLine: String,
                       val remedyShort: String, val reading: Reading)

/** The ported engine core implements this (dasha/varga/ashtakavarga/lattice/synthesis). */
interface AuraEngine {
    fun daily(date: Date): DailyBundle
    // forecast(...), blueprint(...), expanded(...) — same shape as the TS facade.
}

// ── Brand theme (fixed dark; no dynamic color) ───────────────────────────────
@Composable
fun AuraTheme(content: @Composable () -> Unit) {
    val scheme = darkColorScheme(
        background = Tokens.ink, surface = Tokens.card, onBackground = Tokens.text,
        primary = Energy.CRAVE.color, onPrimary = Color(0xFF0B0912), outline = Tokens.line,
    )
    MaterialTheme(colorScheme = scheme, content = content)
}

// ── Aura orb (the signature) ─────────────────────────────────────────────────
@Composable
fun AuraOrb(e1: Color, e2: Color, size: androidx.compose.ui.unit.Dp = 210.dp) {
    val infinite = rememberInfiniteTransition(label = "orb")
    val scale by infinite.animateFloat(
        1f, 1.035f,
        infiniteRepeatable(tween(7000, easing = EaseInOut), RepeatMode.Reverse), label = "breathe")
    Canvas(Modifier.size(size)) {
        val r = this.size.minDimension
        drawCircle(Brush.radialGradient(listOf(e1, Color.Transparent),
            center = Offset(r * 0.32f, r * 0.26f), radius = r * 0.62f))
        drawCircle(Brush.radialGradient(listOf(e2, Color.Transparent),
            center = Offset(r * 0.74f, r * 0.78f), radius = r * 0.62f))
        drawCircle(Brush.radialGradient(listOf(Color.White.copy(alpha = 0.14f), Color.Transparent),
            center = Offset(r * 0.5f, r * 0.44f), radius = r * 0.46f))
    }
    // Apply `scale` via Modifier.scale(scale) at the call site; gate on
    // Settings.Global.ANIMATOR_DURATION_SCALE == 0f for reduce-motion.
}

// ── App shell: Scaffold + NavigationBar (→ NavigationRail on expanded width) ──
enum class Tab(val label: String) { TODAY("Today"), FORECAST("Forecast"), BLUEPRINT("Blueprint") }

@Composable
fun AuraApp(engine: AuraEngine) = AuraTheme {
    var tab by remember { mutableStateOf(Tab.TODAY) }
    Scaffold(
        containerColor = Tokens.ink,
        bottomBar = {
            NavigationBar(containerColor = Tokens.ink.copy(alpha = 0.72f)) {
                Tab.entries.forEach { t ->
                    NavigationBarItem(selected = tab == t, onClick = { tab = t },
                        icon = {}, label = { Text(t.label, fontFamily = Tokens.grotesk, fontSize = 10.sp) })
                }
            }
        }
    ) { pad ->
        Box(Modifier.padding(pad)) {
            when (tab) {
                Tab.TODAY -> TodayScreen(engine.daily(Date()))
                Tab.FORECAST -> Text("Forecast", color = Tokens.text)
                Tab.BLUEPRINT -> Text("Blueprint", color = Tokens.text)
            }
        }
    }
}

// ── Today screen ─────────────────────────────────────────────────────────────
@Composable
fun TodayScreen(daily: DailyBundle) {
    Column(
        Modifier.fillMaxSize().background(Tokens.ink).padding(24.dp),
        horizontalAlignment = Alignment.CenterHorizontally,
        verticalArrangement = Arrangement.spacedBy(22.dp)
    ) {
        AuraOrb(daily.major.color, daily.passing.color)
        Row(Modifier.fillMaxWidth(), horizontalArrangement = Arrangement.SpaceEvenly) {
            BlendCell("Major energy", daily.major)
            BlendCell("Passing through", daily.passing)
        }
        Text(daily.todayLine, fontFamily = Tokens.serif, fontSize = 28.sp, color = Tokens.text,
            textAlign = TextAlign.Center)
        Button(onClick = { /* open reading */ }, modifier = Modifier.fillMaxWidth(),
            shape = RoundedCornerShape(18.dp),
            colors = ButtonDefaults.buttonColors(containerColor = Color.White, contentColor = Color(0xFF0B0912))) {
            Text("Open today’s reading  →", fontFamily = Tokens.grotesk)
        }
    }
}

@Composable
fun BlendCell(label: String, e: Energy) {
    Column(horizontalAlignment = Alignment.CenterHorizontally) {
        Text(label.uppercase(), fontFamily = Tokens.grotesk, fontSize = 10.sp, color = Tokens.textFaint)
        Text(e.label.uppercase(), fontFamily = Tokens.grotesk, fontSize = 15.sp, color = e.color)
    }
}

File: apps/ios/Aura/AuraApp.swift
// aura — iOS (SwiftUI) structural skeleton. Translates the shared design system
// (docs/DESIGN_SYSTEM.md) to native HIG patterns: TabView + NavigationStack, .sheet
// check-in, large serif titles, reduce-motion-aware orb. Compiles against a Swift
// port of the engine core (see docs/PLATFORMS.md, strategy 1); `AuraEngine` is the
// protocol that port implements.

import SwiftUI

// MARK: - Design tokens

enum Theme {
    static let ink      = Color(hex: 0x09080F)
    static let card     = Color.white.opacity(0.035)
    static let line     = Color.white.opacity(0.09)
    static let text     = Color(hex: 0xEDEAF6)
    static let textDim  = Color(hex: 0xB4AFC6)
    static let textFaint = Color(hex: 0x7C7791)

    static let serif   = "InstrumentSerif-Regular"
    static let grotesk = "SpaceGrotesk-Medium"
    static let sans    = "Inter-Regular"

    static let bg = LinearGradient(
        colors: [Color(hex: 0x0A0813), Color(hex: 0x09080F), Color(hex: 0x070610)],
        startPoint: .top, endPoint: .bottom)
}

// MARK: - Domain models (mirror @aura/engine types.ts)

enum Energy: String, CaseIterable {
    case main, feel, fire, mind, grow, love, build, crave, `let`
    var label: String {
        ["main":"Main Character","feel":"Big Feelings","fire":"Fired Up","mind":"Busy Mind",
         "grow":"Green Light","love":"Soft Spot","build":"Heavy Lifting","crave":"Never Enough",
         "let":"Letting Go"][rawValue]!
    }
    var color: Color {
        [ "main":0xFFD070,"feel":0x8FB7FF,"fire":0xFF6E58,"mind":0x5FE0C0,"grow":0x7ED69B,
          "love":0xF49CC9,"build":0x8E93C8,"crave":0xAE8FE6,"let":0xA6ABB8 ][rawValue].map(Color.init(hex:))!
    }
}

struct Reading { let headline, gift, trap, move, watch, remedy: String; let energy, passing: Energy }
struct DailyBundle { let major, passing: Energy; let todayLine, remedyShort: String; let reading: Reading }

/// The ported engine core implements this (dasha/varga/ashtakavarga/lattice/synthesis).
protocol AuraEngine {
    func daily(date: Date) -> DailyBundle
    // forecast(...), blueprint(...), expanded(...) — same shape as the TS facade.
}

// MARK: - Aura orb (the signature)

struct AuraOrb: View {
    let e1: Color, e2: Color
    var size: CGFloat = 210
    @Environment(\.accessibilityReduceMotion) private var reduceMotion
    @State private var breathe = false

    var body: some View {
        ZStack {
            Circle().fill(RadialGradient(colors: [e1, .clear], center: .init(x: 0.32, y: 0.26), startRadius: 0, endRadius: size * 0.62))
            Circle().fill(RadialGradient(colors: [e2, .clear], center: .init(x: 0.74, y: 0.78), startRadius: 0, endRadius: size * 0.62))
            Circle().fill(RadialGradient(colors: [Color.white.opacity(0.14), .clear], center: .init(x: 0.5, y: 0.44), startRadius: 0, endRadius: size * 0.46))
        }
        .frame(width: size, height: size)
        .shadow(color: e2.opacity(0.5), radius: 30)
        .scaleEffect(breathe ? 1.035 : 1.0)
        .onAppear {
            guard !reduceMotion else { return }
            withAnimation(.easeInOut(duration: 7).repeatForever(autoreverses: true)) { breathe = true }
        }
    }
}

// MARK: - App + navigation (TabView → NavigationStack per tab)

@main
struct AuraApp: App {
    var body: some Scene { WindowGroup { RootView() } }
}

struct RootView: View {
    // Injected: the ported engine + a cached chart. Placeholder here.
    var body: some View {
        TabView {
            NavigationStack { TodayView() }
                .tabItem { Label("Today", systemImage: "circle.hexagongrid") }
            NavigationStack { Text("Forecast").foregroundStyle(Theme.text) }
                .tabItem { Label("Forecast", systemImage: "chart.line.uptrend.xyaxis") }
            NavigationStack { Text("Blueprint").foregroundStyle(Theme.text) }
                .tabItem { Label("Blueprint", systemImage: "sparkles") }
        }
        .tint(Energy.smokeTint)
        .background(Theme.bg.ignoresSafeArea())
    }
}

// MARK: - Today screen

struct TodayView: View {
    @State private var showCheckin = false
    // Placeholder bundle; real app pulls from AuraEngine.daily(date:).
    private let major = Energy.build, passing = Energy.crave

    var body: some View {
        ScrollView {
            VStack(spacing: 22) {
                AuraOrb(e1: major.color, e2: passing.color)
                HStack(spacing: 0) {
                    blendCell("Major energy", major)
                    Divider().frame(height: 40).overlay(Theme.line)
                    blendCell("Passing through", passing)
                }
                Text("Build the brick in front of you.")
                    .font(.custom(Theme.serif, size: 30)).foregroundStyle(Theme.text)
                    .multilineTextAlignment(.center).padding(.top, 4)
                remedyPill
                Button { /* open reading */ } label: {
                    Text("Open today’s reading  →").frame(maxWidth: .infinity)
                }
                .buttonStyle(AuraPrimaryButton())
            }
            .padding(24)
        }
        .background(Theme.bg.ignoresSafeArea())
        .navigationTitle("aura")
        .sheet(isPresented: $showCheckin) { Text("Check-in sheet").presentationDetents([.medium]) }
    }

    private func blendCell(_ label: String, _ e: Energy) -> some View {
        VStack(spacing: 6) {
            Text(label.uppercased()).font(.custom(Theme.grotesk, size: 10)).tracking(2).foregroundStyle(Theme.textFaint)
            Text(e.label.uppercased()).font(.custom(Theme.grotesk, size: 15)).foregroundStyle(e.color)
        }.frame(maxWidth: .infinity)
    }

    private var remedyPill: some View {
        HStack(spacing: 11) {
            Circle().strokeBorder(Energy.crave.color, lineWidth: 1.6).frame(width: 22, height: 22)
            VStack(alignment: .leading, spacing: 2) {
                Text("TODAY’S REMEDY").font(.custom(Theme.grotesk, size: 9)).tracking(2).foregroundStyle(Theme.textFaint)
                Text("Fix your sleep window — up early, down early").font(.custom(Theme.sans, size: 14)).foregroundStyle(Theme.text)
            }
            Spacer()
        }
        .padding(14).background(Theme.card).clipShape(RoundedRectangle(cornerRadius: 16))
        .overlay(RoundedRectangle(cornerRadius: 16).stroke(Theme.line))
        .onTapGesture { showCheckin = true }
    }
}

struct AuraPrimaryButton: ButtonStyle {
    func makeBody(configuration: Configuration) -> some View {
        configuration.label
            .font(.custom(Theme.grotesk, size: 15)).foregroundStyle(Color(hex: 0x0B0912))
            .padding(.vertical, 17)
            .background(LinearGradient(colors: [.white, Color(hex: 0xE4E0F0)], startPoint: .top, endPoint: .bottom))
            .clipShape(RoundedRectangle(cornerRadius: 18))
            .opacity(configuration.isPressed ? 0.85 : 1)
    }
}

// MARK: - Helpers

extension Color {
    init(hex: Int) {
        self.init(.sRGB,
                  red: Double((hex >> 16) & 0xFF) / 255,
                  green: Double((hex >> 8) & 0xFF) / 255,
                  blue: Double(hex & 0xFF) / 255)
    }
}
extension Energy { static let smokeTint = Color(hex: 0xAE8FE6) }
